File: blockchain-mobile/lib/models/order.dart
import 'package:blockchain_mobile/models/Product.dart';
import 'package:blockchain_mobile/models/enums/e_received_status.dart';

import 'enums/e_process_receive_product.dart';
import 'enums/e_user_confirm.dart';

class Order {
  int id;
  DateTime createDate;
  EReceivedStatus statusReceived;
  EUserConfirm userConfirm;
  String qrCode;
  dynamic discountCode;
  double totalAmount;
  String firstName;
  String lastName;
  String email;
  String phoneNumber;
  String? address;
  List<OrderDetail> orderDetails;
  bool consignment;

  static String? errorQrScanMessage = "This is an Order qr Code";

  Order({
    required this.id,
    required this.createDate,
    required this.statusReceived,
    required this.qrCode,
    required this.discountCode,
    required this.totalAmount,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phoneNumber,
    required this.address,
    required this.userConfirm,
    required this.orderDetails,
    required this.consignment,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    String statusReceived = json['statusReceived'];
    return Order(
      id: json['id'],
      createDate: DateTime.parse(json['createDate']).toLocal(),
      statusReceived: EReceivedStatus.values
          .firstWhere((element) => element.name == statusReceived),
      qrCode: json['qrCode'],
      discountCode: json['discountCode'],
      totalAmount: json['totalAmount'].toDouble(),
      firstName: json['firstName'],
      lastName: json['lastName'],
      email: json['email'],
      phoneNumber: json['phoneNumber'],
      address: json['address'] ?? "",
      orderDetails: List<OrderDetail>.from(
          json['orderDetails'].map((x) => OrderDetail.fromJson(x))),
      userConfirm: EUserConfirm.values.firstWhere(
          (element) => element.name == json['userConfirm'],
          orElse: () => EUserConfirm.NULL),
      consignment: json['consignment'],
    );
  }

  @override
  String toString() {
    return 'Order{id: $id, createDate: $createDate, statusReceived: $statusReceived, qrCode: $qrCode, discountCode: $discountCode, totalAmount: $totalAmount, firstName: $firstName, lastName: $lastName, email: $email, phoneNumber: $phoneNumber, address: $address, orderDetails: $orderDetails, consignment: $consignment}';
  }
}

class OrderDetail {
  int id;
  Product product;
  int quantity;
  double price;
  double amount;
  EProcessReceiveProduct processReceiveProduct;

  OrderDetail({
    required this.id,
    required this.product,
    required this.quantity,
    required this.price,
    required this.amount,
    required this.processReceiveProduct,
  });

  factory OrderDetail.fromJson(Map<String, dynamic> json) {
    var processReceiveProduct = json['processReceiveProduct'];
    return OrderDetail(
      id: json['id'],
      product: Product.fromJson(json['product']),
      quantity: json['quantity'],
      price: json['price'].toDouble(),
      amount: json['amount'].toDouble(),
      processReceiveProduct: EProcessReceiveProduct.values
          .firstWhere((element) => element.name == processReceiveProduct),
    );
  }
  @override
  String toString() {
    return 'OrderDetail{id: $id, product: $product, quantity: $quantity, price: $price, amount: $amount, processReceiveProduct: $processReceiveProduct}';
  }
}

File: blockchain-server/src/main/java/com/server/blockchainserver/models/shopping_model/Order.java
package com.server.blockchainserver.models.shopping_model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.server.blockchainserver.models.enums.EReceivedStatus;
import com.server.blockchainserver.models.enums.EUserConfirm;
import com.server.blockchainserver.models.user_model.User;
import jakarta.persistence.*;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

@Entity
@Table(name = "orders")
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id")
    @JsonIgnore
    private User user;
    private Instant createDate;

    @Enumerated(EnumType.STRING)
    @Column(length = 100)
    private EReceivedStatus statusReceived;
@Column(name = "qr_code")
    private String qrCode;
    private String discountCode;
    private Double percentageDiscount;
    private BigDecimal totalAmount;
    private String firstName;
    private String lastName;
    private String email;
    private String phoneNumber;
    private String address;
    private boolean isConsignment;
    @Enumerated(EnumType.STRING)
    @Column(length = 100)
    private EUserConfirm userConfirm;
    @OneToMany(mappedBy = "order", fetch = FetchType.LAZY)
    private List<OrderDetail> orderDetails;

    public Order() {
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Instant getCreateDate() {
        return createDate;
    }

    public void setCreateDate(Instant createDate) {
        this.createDate = createDate;
    }

    public BigDecimal getTotalAmount() {
        return totalAmount;
    }

    public void setTotalAmount(BigDecimal totalAmount) {
        this.totalAmount = totalAmount;
    }

    public String getDiscountCode() {
        return discountCode;
    }

    public void setDiscountCode(String discountCode) {
        this.discountCode = discountCode;
    }

    public Double getPercentageDiscount() {
        return percentageDiscount;
    }

    public void setPercentageDiscount(Double percentageDiscount) {
        this.percentageDiscount = percentageDiscount;
    }

    public List<OrderDetail> getOrderDetails() {
        return orderDetails;
    }

    public void setOrderDetails(List<OrderDetail> orderDetails) {
        this.orderDetails = orderDetails;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public EReceivedStatus getStatusReceived() {
        return statusReceived;
    }

    public void setStatusReceived(EReceivedStatus statusReceived) {
        this.statusReceived = statusReceived;
    }

    public String getQrCode() {
        return qrCode;
    }

    public void setQrCode(String qrCode) {
        this.qrCode = qrCode;
    }

    public boolean isConsignment() {
        return isConsignment;
    }

    public void setConsignment(boolean consignment) {
        isConsignment = consignment;
    }

    public EUserConfirm getUserConfirm() {
        return userConfirm;
    }

    public void setUserConfirm(EUserConfirm userConfirm) {
        this.userConfirm = userConfirm;
    }
}


